### DISCUSSION TITLE
Demo: ElkJS `partitioning` setting to group nodes by type & ensure ordering. Example of visualizing complex networks

### DISCUSSION THREAD

**Discussion by cliffordfajardo** [xyflow/xyflow#3355] [Layouting]
### Background Context
I needed to visualize a bunch of nodes and lay them out in a particular order from left to right **_and always guarantee they are in a specific order._**

Conceptually, what I wanted was to group certain items by column/partition kind of like a Sankey diagram
Essentially I needed a way to tell Elkjs that some set of nodes belong to a certain column or partition like this:
<img height="300" src="https://github.com/reaviz/reaflow/assets/6743796/1c91592a-b64f-4fc0-af51-13a697a49e7c" />

😢 It took me weeks to figure out how to do this so I'd figure other folks might find this helpful

## Caveats / Warning ⚠️ 
1. In the example below I am currently using a different node visualization library called [reaflow](https://github.com/reaviz/reaflow), but i think a lot of this can transfer to reactflow if your using it with elkjs ([reaflow + elkjs example docs](https://reactflow.dev/docs/examples/layout/elkjs/))

2. If you've achieved this without ElkJS please leave a comment or example below 🙏 

## How this works
> Code is based off of reaflow library, but you can pass these settings to `[elkOptions` with reactflow](https://reactflow.dev/docs/examples/layout/elkjs/)

You need to activate "partioning" inside of `layoutOptions` to activate this settings globally for your graph:
```ts
<Canvas 
   nodes={nodes}
   edges={edges}
   layoutOptions={{
    'elk.algorithm': 'layered',              // < ------------✅ 
    'elk.partitioning.activate': 'true',  // < ------------✅ 
  }}
/>
```
Then in your node objects you need to pass `layoutOptions` property with the ordering like this:
```ts
const nodes = [
  {
    id: 'node01',
    text: 'node01',
    className: 'blue-green',
    layoutOptions: {
      'partitioning.partition': 1, / < ------------✅ 
    },
  },
  {
    id: 'node02',
    text: 'node02',
    className: 'blue-green',
    layoutOptions: {
      'partitioning.partition':  2, // < ------------✅ 
    },
  },
]
```

## Example Code / Visualization
I have a simpler example of the layout above in this stackblitz code sandbox:
https://stackblitz.com/edit/react-ts-mgkmmq?file=data%2Fedges.ts,data%2Fnodes.ts,App.tsx,data%2Findex.ts
<img height="300" src="https://github.com/reaviz/reaflow/assets/6743796/e189465b-1eb8-48da-810d-4ca7f15d7254" />

**Comment by moklick** [comment]
Hey @cliffordfajardo thanks for sharing! This is super valuable since ELKjs is really hard to configure.

**Comment by blqke** [comment]
You just saved me hours
